go/marshal/decode.go

```diff
@@ -0,0 +1,227 @@
+type OverflowError struct {
+	Value types.Number
+	Type  reflect.Type // type of Go value it could not be assigned to
+}
+
+func (e *OverflowError) Error() string {
+	return fmt.Sprintf("Cannot unmarshal %g into Go value of type %s", e.Value, e.Type)
+}
+
+type decoderFunc func(v types.Value, rv reflect.Value)
+
+func typeDecoder(t reflect.Type) decoderFunc {
+	switch t.Kind() {
+	case reflect.Bool:
+		return boolDecoder
+	case reflect.Float32, reflect.Float64:
+		return floatDecoder
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return intDecoder
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return uintDecoder
+	case reflect.String:
+		return stringDecoder
+	case reflect.Struct, reflect.Interface:
+		return structDecoder(t)
+	default:
+		panic(&UnsupportedTypeError{Type: t})
+	}
+}
+func boolDecoder(v types.Value, rv reflect.Value) {
+	if b, ok := v.(types.Bool); ok {
+		rv.SetBool(bool(b))
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func stringDecoder(v types.Value, rv reflect.Value) {
+	if s, ok := v.(types.String); ok {
+		rv.SetString(string(s))
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func floatDecoder(v types.Value, rv reflect.Value) {
+	if n, ok := v.(types.Number); ok {
+		rv.SetFloat(float64(n))
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func intDecoder(v types.Value, rv reflect.Value) {
+	if n, ok := v.(types.Number); ok {
+		i := int64(n)
+		if rv.OverflowInt(i) {
+			panic(&OverflowError{n, rv.Type()})
+		}
+		rv.SetInt(i)
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func uintDecoder(v types.Value, rv reflect.Value) {
+	if n, ok := v.(types.Number); ok {
+		u := uint64(n)
+		if rv.OverflowUint(u) {
+			panic(&OverflowError{n, rv.Type()})
+		}
+		rv.SetUint(u)
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+var structDecoderCache struct {
+	sync.RWMutex
+	m map[reflect.Type]decoderFunc
+}
+
+type decField struct {
+	name    string
+	decoder decoderFunc
+	index   int
+}
+
+func structDecoder(t reflect.Type) decoderFunc {
+	// TODO: Is this handling wrong noms type?
+	if t.Implements(nomsValueType) {
+		return nomsValueDecoder
+	}
+
+	structDecoderCache.RLock()
+	d := structDecoderCache.m[t]
+	structDecoderCache.RUnlock()
+	if d != nil {
+		return d
+	}
+
+	name := t.Name()
+	fields := make([]decField, 0, t.NumField())
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		validateField(f, t)
+
+		tags := f.Tag.Get("noms")
+		if tags == "-" {
+			continue
+		}
+
+		fields = append(fields, decField{
+			name:    getFieldName(tags, f),
+			decoder: typeDecoder(f.Type),
+			index:   i,
+		})
+	}
+
+	d = func(v types.Value, rv reflect.Value) {
+		s := v.(types.Struct)
+		if s.Type().Desc.(types.StructDesc).Name != name {
```

Comment:
Noms structs can have empty names, right? I think this means that you can't marshal/unmarshal nameless Noms struct types, and I don't know that we want that restriction